src/pages/experiments/[experimentId]/data-exploration/components/heatmap/HeatmapPlot.jsx

```diff
@@ -121,83 +122,148 @@ const HeatmapPlot = (props) => {
+        // Union of allCellsInSets and cellSet
+        allCellsInSets = new Set([...allCellsInSets, ...cellSet]);
+      });
+    });
 
-    // Find the `groupBy` root node.
-    const rootNode = hierarchy.filter((clusters) => clusters.key === groupBy);
+    // Only work with non-hidden cells.
+    const shownCells = new Set([...allCellsInSets].filter(x => !hiddenCellIds.has(x)));
 
-    if (!rootNode.length) {
-      return [];
+    return shownCells;
+  }
+
+  const getCellSetNamesDict = (rootNodes) => {
+    const cellSetNamesByGroupByKey = {};
+    rootNodes.forEach((currentRootNode) => cellSetNamesByGroupByKey[currentRootNode.key] = currentRootNode.children);
+
+    return cellSetNamesByGroupByKey
+  };
+
+  const downsampled = (cellIds) => {
+    // If sample is within maxCells's limit then return original array
+    if (cellIds.length <= maxCells) {
+      return cellIds;
     }
 
-    const { children } = rootNode[0];
+    let downsampledCellIds = [];
 
-    // Iterate over each child node.
-    children.forEach(({ key }) => {
-      // Only work with non-hidden cells.
-      const shownCells = Array.from(
-        properties[key].cellIds,
-      ).filter(
-        (id) => !hiddenCellIds.has(id),
-      );
+    // Take maxCells amount of cells, picking each one randomly
+    for (let i = 0; i < maxCells; i++) {
+      var randomIndex = Math.floor(Math.random() * cellIds.length);
+      const element = cellIds.splice(randomIndex, 1);
 
-      total += shownCells.length;
+      downsampledCellIds.push(element[0]);
+    }
 
-      groupedCells[key] = shownCells;
-    });
+    return downsampledCellIds;
+  };
+
+  const downsampleAndSort = (groupByTracks) => {
+    // Find the `groupBy` root nodes.
 
-    // If we collected less than `max` number of cells, let's go with that.
-    const finalSampleSize = Math.min(total, maxCells);
+    // About the filtering: If we have failed to find some of the groupbys information, 
+    // then ignore those (this is useful for groupbys that sometimes dont show up, like 'samples')
+    let groupByRootNodes =
+      groupByTracks
+        .map((groupByKey) => {
+          return hierarchy.find((cluster) => (cluster.key === groupByKey))
+        })
+        .filter((track => track !== undefined));
 
-    if (total === 0) {
+
+    if (!groupByRootNodes.length) {
       return [];
     }
 
-    // Create a sample of cells to display.
-    const sample = [];
+    // Transform the groupByRootNodes element into a dictionary/object to make its searches by key faster 
+    // (this is important since it will be used in the sorting comparator) 
+    const cellSetNamesByGroupByKey = getCellSetNamesDict(groupByRootNodes);
+
+    // Get all cellIds from cells that will show up (so those that aren't hidden and are in an enabled groupBy)
+    const cellIdsSet = getAllRelevantCellIds(cellSetNamesByGroupByKey);
+    const cellIdsArray = Array.from(cellIdsSet);
+
+    const cellIdsSample = downsampled(cellIdsArray);
```

Comment:
I thought the idea here was that cells would be sampled randomly, but in a way such that each bottom-level grouping would have their proportions preserved, not just a literal random sampling of all cells. If you sample all cells randomly, you will have most of the cells from the biggest clusters, and almost none from the smaller ones, which is exactly what we want to avoid. Maybe I'm misunderstanding.

Reply:
I am not sampling completely randomly, first I am picking only the cells that belong to one or more of the enabled group bys.

That was what I interpreted could be done to reduce the impact on how representative the downsample is.